### PR TITLE
PDL for parametric AC-OPF: paper-faithful implementation with analytical dual

### PR DESCRIPTION
**Pull request by xkhainguyen**
## Summary

Implementation of Primal-Dual Learning (PDL) (Park & Van Hentenryck, AAAI 2023) for parametric AC-OPF, with several fixes and extensions over a naive baseline.

### Fixes to the base implementation

| Fix | Effect |
|-----|--------|
| Paper-faithful augmented Lagrangian loss (raw g, h from `BNK.constraints!`) | Correct Lagrangian gradient; sign of μᵀg term pushes primal toward feasibility |
| Clamp inequality multipliers to μ ≥ 0 | KKT sign correctness |
| `BoundedOutput` sigmoid head | `max_bound_viol ≡ 0` by construction; no dead gradient zone (vs hardsigmoid) |
| `FixRefBus` layer | Reference-bus angle fixed architecturally; GPU-safe |
| Analytical primal loss | Eliminates dual tracking gap — see below |

### Analytical dual correction

The dual network lags the true multipliers by orders of magnitude at high ρ. Instead of using `λ̂(θ)` directly, we apply one ALM update analytically at every primal gradient step:

```
μ_eff = clamp(μ̂(θ) + ρ·g(ŷ),  0,  max_dual)
λ_eff = clamp(λ̂(θ) + ρ·h(ŷ), −max_dual, max_dual)
```

Gradient flows through g(ŷ) and h(ŷ) — not through the dual net — giving the correct ρg² and ρh² penalty terms without tracking lag. The dual network still trains in parallel, providing an improving warm-start base over time.

### API

Both options live in `ALMMethod` (defaults shown):

```julia
method = ALMMethod(;
    batch_model         = bm,
    num_equal           = num_equal,
    use_analytical_dual = true,   # recommended; false = use dual net output directly
    use_dual_learning   = true,   # false = freeze dual network (ablation only)
    ρmax = 1e4, max_dual = 1e6, τ = 0.8, α = 2.0,
)
train!(method, trainer, data; K=100, L_primal=2500, L_dual=2500, warmup_epochs=25000)
```

### Results on case57 (5000 held-out test samples)

| Config | max_eq | max_ineq |
|--------|--------|----------|
| `use_analytical_dual=true`, ρmax=1e4, max_dual=1e6 | **1.165** | 0.000 |
| `use_analytical_dual=false` | ~1.25 | 0.000 |
| max_dual=1e4 (any ρ schedule) | ~1.80 | 0.000 |

Variable bounds and inequalities are satisfied exactly in all runs.

## Test plan
- [ ] `julia --project=. -e 'using Pkg; Pkg.test()'` passes on CPU
- [ ] `BNK_TEST_CUDA=1 julia --project=. -e 'using Pkg; Pkg.test()'` passes on GPU
- [ ] `julia --project=. examples/case57_train.jl` runs to completion

### CODE REVIEW

**Comment by andrewrosemberg** [comment]
Amazing!! Let us know when you are ready for a review

**Comment by xkhainguyen** [comment]
Hi @andrewrosemberg,

I wanted to mention that this is my personal attempt to reproduce the paper's approach. I tried to fix the branch's implementation and make it work on the 57-bus case. The results are reported above. To be honest, I was hoping for even better results, but it may be a limitation of the approach. Feel free to review and even improve or tune it further.

Many thanks!
Khai

File: src/L2OALM.jl [hunk 1]
```diff
@@ -181,33 +273,60 @@ from current dual predictions for the batch model `bm` under parameters `Θ`.
...
+        μ = max.(dual_hat[1:end-num_equal, :], zero(eltype(dual_hat)))
+        λ = dual_hat[end-num_equal+1:end, :]
+
+        # Compute violations from gh directly — avoids a second BNK.constraints! call
+        # through all_violations! (which shares bm.cons_out and corrupts the rrule view).
+        Vc_ineq = max.(g, zero(eltype(g)))
+        Vc_eq   = abs.(h)
+        Vc      = vcat(Vc_ineq, Vc_eq)
+
+        lagrangian_term = (sum(μ .* g) + sum(λ .* h)) / num_s
```
Comment (andrewrosemberg) [inline src/L2OALM.jl:318]: @ivanightingale , I believe this was one of the issues we discussed correct?

The fact that we are not correcting the sign after multiplying dual by constraint value.
In my implementation, I just decided to use the violations instead of the function values, but I see the issue with that too

Comment (ivanightingale) [reply src/L2OALM.jl:318]: The current PR is faithful to the PDL paper, but like we discussed the paper didn't get ALM exactly right.

Here is the proper augmented Lagrangian (primal loss) from
> Rockafellar, R. Tyrrell. "Augmented Lagrange multiplier functions and duality in nonconvex programming." SIAM Journal on Control 12.2 (1974): 268-285.

and a derivation:

Screenshot 2026-06-01 at 12 48 16 PM

More modern implementations like `ALGENCAN` usually use an equivalent primal objective. So strictly speaking, the "augmented Lagrangian" is not really just "the Lagrangian plus a penalty term".

Comment (ivanightingale) [reply src/L2OALM.jl:318]: In fact I have implemented ALGENCAN, LANCELOT, and a version of ALM that uses the augmented Lagrangian from the PDL paper, all under a same framework. The first two converge, but the last one doesn't. I'm talking about optimization algorithms here.

So from a "motivated by the optimization algorithm" perspective, one would want to try swapping in a proper augmented Lagrangian as the primal net loss in PDL. I have tried doing this, but it didn't change the convergence of training too much.

File: src/L2OALM.jl [hunk 2]
```diff
@@ -220,12 +339,14 @@ This function computes the maximum violation, mean violations, mean objectives,
...
-    max_violation = maximum([s.max_violation for s in batch_states])
+    # Use mean of mean_violations (not max-of-max) as the ρ-update signal.
+    # max-of-max is dominated by noisy hard batches and causes ρ to escalate
+    # before violations have genuinely plateaued, collapsing generators.
     mean_violations = mean([s.mean_violations for s in batch_states])
     mean_objs = mean([s.mean_objs for s in batch_states])
     mean_loss = mean([s.total_loss for s in batch_states])
     return (;
-        max_violation = max_violation,
+        max_violation = mean_violations,   # proxy for ρ criterion: bump only when mean stalls
```
Comment (ivanightingale) [inline src/L2OALM.jl:349]: The paper uses the max-of-max, but maybe mean-of-max is better.

File: src/L2OALM.jl (hunk 1 above)
Comment (xkhainguyen) [reply src/L2OALM.jl:318]: Thanks, this is insightful.

In case you guys have the new versions of ALM. Could you also benchmark them to see if they can improve my results?

Comment (andrewrosemberg) [reply src/L2OALM.jl:318]: could be nice to have a way to allow users to choose between these different options! Any ideas on how we design the API @xkhainguyen @ivanightingale ? after we implement, I can run all on comparable hardware and cases to benchmark.

My implementation worked well up to case 6K rte but took a while to converge

File: src/L2OALM.jl (hunk 2 above)
Comment (xkhainguyen) [reply src/L2OALM.jl:349]: Oh, I didn’t notice. I’ve always felt that the field is missing a standardized set of metrics that every paper should report.

@andrewrosemberg @ivanightingale I'd be happy to contribute to a community effort around this.

File: src/L2OALM.jl (hunk 1 above)
Comment (xkhainguyen) [reply src/L2OALM.jl:318]: I guess we can just split cases / approaches within function [single_train_step!](https://github.com/xkhainguyen/L2OALM.jl/blob/b4333c462c6740fd9cdd6d1ceca1d95c97ba0d5c/src/L2OALM.jl#L442)

and then choose options here
```
method = ALMMethod(;
    batch_model         = bm,
    num_equal           = num_equal,
    use_analytical_dual = true,   # recommended; false = use dual net output directly
    use_dual_learning   = true,   # false = freeze dual network (ablation only)
    ρmax = 1e4, max_dual = 1e6, τ = 0.8, α = 2.0,
)
train!(method, trainer, data; K=100, L_primal=2500, L_dual=2500, warmup_epochs=25000)
```

Feel free to add your approaches there

File: src/L2OALM.jl (hunk 2 above)
Comment (andrewrosemberg) [reply src/L2OALM.jl:349]: We started this organization to do just that. We can explore here and then move the evaluation metrics to a specific repo. We were creating one repo for each big method (still missing a few like deepmind cosmos ) and we were going to train and benchmark on a public dataset https://huggingface.co/datasets/LearningToOptimize/Parametric-Optimization-Problems)

@xkhainguyen happy for us all to have a call and plan :)
